go/libraries/doltcore/sqle/index/z_encoding.go

```diff
@@ -45,45 +47,116 @@ func UnLexFloat(b uint64) float64 {
+	for i := 0; i < 5; i++ {
+		x = (x | (x << shifts[i])) & masks[i]
+		y = (y | (y << shifts[i])) & masks[i]
 	}
-	return res
+	return x | (y << 1)
 }
 
-// UnZValue takes a [16]byte Z-Value and converts it back to a sql.Point
-func UnZValue(z [16]byte) types.Point {
-	var x, y uint64
-	for i := 15; i >= 0; i-- {
-		zv := uint64(z[i])
-		for j := 3; j >= 0; j-- {
-			y |= (zv & 1) << (63 - (4*i + j))
-			zv >>= 1
+// ZValue takes a Point and interleaves the bits into a [2]uint64
+// It will put the bits in this order: x_0, y_0, x_1, y_1 ... x_63, Y_63
+func ZValue(p types.Point) (z [2]uint64) {
+	xLex, yLex := LexFloat(p.X), LexFloat(p.Y)
+	z[0], z[1] = InterleaveUInt64(xLex>>32, yLex>>32), InterleaveUInt64(xLex&0xFFFFFFFF, yLex&0xFFFFFFFF)
+	return
+}
 
-			x |= (zv & 1) << (63 - (4*i + j))
-			zv >>= 1
-		}
+// UnInterleaveUint64 splits up the bits of the uint64 z into two uint64s
+// The first 32 bits of x and y must be 0.
+// Example:
+// abcd efgh ijkl mnop abcd efgh ijkl mnop abcd efgh ijkl mnop abcd efgh ijkl mnop 0x5555555555555555
+// 0b0d 0f0h 0j0l 0n0p 0b0d 0f0h 0j0l 0n0p 0b0d 0f0h 0j0l 0n0p 0b0d 0f0h 0j0l 0n0p x | x >> 1
+// 0bbd dffh hjjl lnnp pbbd dffh hjjl lnnp pbbd dffh hjjl lnnp pnbd dffh hjjl lnnp 0x3333333333333333
+// 00bd 00fh 00jl 00np 00bd 00fh 00jl 00np 00bd 00fh 00jl 00np 00bd 00fh 00jl 00np x | x >> 2
+// 0000 bdfh fhjl jlnp npbd bdfh fhjl jlnp npdb bdfh fhjl jlnp npdb bdfh fhjl jlnp 0x0F0F0F0F0F0F0F0F
+// 0000 bdfh 0000 jlnp 0000 bdfh 0000 jlnp 0000 bdfh 0000 jlnp 0000 bdfh 0000 jlnp x | x >> 4
+// 0000 bdfh bdfh jlnp jlnp bdfh bdfh jlnp jlnp bdfh bdfh jlnp jlnp bdfh bdfh jlnp 0x00FF00FF00FF00FF
+// 0000 0000 bdfh jlnp 0000 0000 bdfh jlnp 0000 0000 bdfh jlnp 0000 0000 bdfh jlnp x | x >> 8
+// 0000 0000 0000 0000 bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp 0x0000FFFF0000FFFF
+// 0000 0000 0000 0000 bdfh jlnp bdfh jlnp 0000 0000 0000 0000 bdfh jlnp bdfh jlnp x | x >> 16
+// 0000 0000 0000 0000 bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp 0x00000000FFFFFFFF
+// 0000 0000 0000 0000 0000 0000 0000 0000 bdfh jlnp bdfh jlnp bdfh jlnp bdfh jlnp
+func UnInterleaveUint64(z uint64) (x, y uint64) {
+	x, y = z, z>>1
+	for i := 4; i >= 0; i-- {
+		x &= masks[i]
+		x |= x >> shifts[i]
+		y &= masks[i]
+		y |= y >> shifts[i]
 	}
-	xf := UnLexFloat(x)
-	yf := UnLexFloat(y)
+	x &= 0xFFFFFFFF
+	y &= 0xFFFFFFFF
+	return
+}
+
+// UnZValue takes a [2]uint64 Z-Value and converts it back to a sql.Point
+func UnZValue(z [2]uint64) types.Point {
+	xl, yl := UnInterleaveUint64(z[0])
+	xr, yr := UnInterleaveUint64(z[1])
+	xf := UnLexFloat((xl << 32) | xr)
+	yf := UnLexFloat((yl << 32) | yr)
 	return types.Point{X: xf, Y: yf}
 }
 
 func ZSort(points []types.Point) []types.Point {
 	sort.Slice(points, func(i, j int) bool {
 		zi, zj := ZValue(points[i]), ZValue(points[j])
-		return bytes.Compare(zi[:], zj[:]) < 0
+		if zi[0] == zj[0] {
+			return zi[1] < zj[1]
+		}
+		return zi[0] < zj[0]
```

Comment:
can make this condition branchless
```suggestion
		return zi[0] < zj[0] || (zi[0] == zj[0] && zi[1] < zi[1])
```